src/main/java/net/querz/nbt/CompoundTag.java

```diff
@@ -293,6 +289,13 @@ public boolean equals(Object other) {
 		return true;
 	}
 
+	@Override
+	public int hashCode() {
+		int keyHash = Objects.hash(getValue().keySet().toArray());
```

Comment:
Why do you create here and in other cases an array first?

Reply:
>calculate the hash code of just the key set

Isn't that exactly the wanted behavior, or rather of the map?

While thinking about it maybe it would be even easier to implement equals and hashCode for the `Tag` base class using [`Objects#deepEquals`](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#deepEquals-java.lang.Object-java.lang.Object-) and build or own "`Objects.deepHashCode`" which uses the respective `Objects.hashCode` / `Arrays.deepHashCode` / `Arrays.hashCode`.

With the `toArray` approach there are at least two problems:
- It imposes and order on the set elements; therefore two (key-)sets with the same elements but in different order might have differing hash codes.
- It is rather inefficient because every time a new array has to be allocated, see [`Set.toArray()`](https://docs.oracle.com/javase/8/docs/api/java/util/Set.html#toArray--)

Reply:
Or for the array tag types maybe just overwrite the `hashCode` method?

Reply:
Nevermind, you seem to have changed it already